### ISSUE TITLE
Explicitly map ValueType to specific Intent enums and record value interfaces

### ISSUE THREAD

**Issue by npepinpe** [camunda-cloud/zeebe#7587]
**Description**

One thing that came out of #7545 is that we duplicate, again, mapping `ValueType` constants to specific `Intent` implementation enums and `RecordValue` interfaces/implementation classes. Now, we can't avoid having to map each time explicitly for each implementation, but we can avoid having to map in multiple places each `ValueType` to its corresponding `Intent` enum and `RecordValue` interface. I would propose the following (abridged):

```java
public enum ValueType {
  ERROR((short) 0, ErrorRecordValue.class, ErrorIntent.class),
  INCIDENT((short)1, IncidentRecordValue.class, IncidentIntent.class);

  private final short value;
  private final Class<? extends RecordValue> valueClass;
  private final Class<? extends Intent> intentEnum;

  private ValueType(
    final short value, 
    final Class<? extends RecordValue> valueClass, 
    final Class<? extends Intent> intentEnum) {
    this.value = value;
    this.valueClass = valueClass;
    this.intentEnum = intentEnum;
  }

  public short getValue() {
    return value;
  }

  public Class<? extends RecordValue> getValueClass() {
    return valueClass;
  }

  public Class<? extends Intent> getIntentEnum() {
    return intentEnum;
  }
}
```

The `intentEnum` mapping would reduce the maintenance cost of maintaining two separate mappings in the `Intent` interface, which is what we currently do, to map serialized intent values back to their original values.

The `valueClass` mapping would allow consumers, which typically want to deserialize a value based on its value type into some concrete implementation class, to verify that the implementation type corresponds to a valid expect interface. So any implementation which does deserialization can have a smoke test to ensure that its own mapping of the `ValueType` to the implementation class matches the expected types as defined by the protocol. This would provide a bit more safety than via the implicit class cast which is done in most places, and allow for a more explicit error message, e.g. `Expected to deserialize record with value type VARIABLE into ErrorRecordValueImpl, but it is not extending the expected base type of VariableRecordValue`.

The main downside here is that `ValueType` is a generated enum at the moment, so we can't quite modify it. We have two options:

1. Extract it and maintain it manually. This implies a bit more work on the SBE side, since encoding/decoding will not be via the enum but via the short value directly.
1. Have this mapping into a separate class, e.g. `ValueTypes` which provides the methods above probably delegating to an internal `EnumMap` or some such.

**Comment by saig0** [comment]
@npepinpe probably, I miss some context. Please explain where and how this class would be used.

**Comment by npepinpe** [comment]
Possibly, I'll explain it and once clarified update the original issue :smile: 

TL;DR - I would like to reduce a bit the boilerplate when adding new records, and make the `ValueType` to `Intent` and `RecordValue` more explicit. I'm fully aware that the proposal only is low impact, so I wouldn't be pushing for us to do it anytime soon to be honest. I'd still like to bring it up, as I think it's not super clear for new comers what they have to do when adding new record types, or how these three types relate to each other, so anything to help there is a win.

We have a `ValueType` enum, which, from what I can tell, is primarily used for polymorphic deserialization of the `Record<T>` interface, primarily the `T` in there, i.e. the record's `value`; this is how we know to deserialize the record's value into, say, a `VariableRecord`, or an `ErrorRecord`. We also use it for the deserialization of the `intent` value in a concrete enum type, e.g. `ErrorIntent`, `IncidentIntent`, etc. 

Currently, we do this via three mappings: in the `Intent` interface, we have two methods which map a `ValueType` to the concrete `Intent` enum, via a `switch` construct. One method uses the intent's `value` property and the `ValueType`, and the other uses a `ValueType` and the intent's name (note: that second one isn't actually used anywhere :smile:). Then we have another mapping in the `TypedEventRegistry`, in the engine. This maps a `ValueType` to a cached instance of a record, i.e. `ValueType.VARIABLE` maps to a `VariableRecord` instance. This is then used to deserialize the record from the log stream into this object.

This means, when you add a new record type, you need to:

1. Add a new entry to the `ValueType` enum
1. Add a new `Intent` class
1. Add a new `RecordValue` interface
1. Add a new `UnifiedRecordValue` implementation
1. Register the new value type in the two intent mappings
1. Register the new value type in the `TypedEventRegistry`

When we add a new deserialization and implementation of the record values, we now have an additional place where we need to map `ValueType` to an implementation class, and unfortunately to an intent. We do this by implementing custom [TypeIdResolver](https://fasterxml.github.io/jackson-databind/javadoc/2.7/com/fasterxml/jackson/databind/jsontype/TypeIdResolver.html). Jackson has this nice property where you can say that, for example, the value's concrete type is resolved from its sibling property the `valueType`. Your `TypeIdResolver` then receives the value type (e.g. `ValueType.INCIDENT`) and must provide a class into which the `value` will be deserialized. It works the same for the intent - it must provide an `Intent` sub class.

This means we now also have to add to the above:

1. Adding a new abstract class that implements the new interface (but nothing to implement) in the `protocol-jackson` module. It may need to be customized if it contains further nested types where we would need to add deserialization notices to Jackson.
1. Registering the new value type in the `ValueTypes` mapping of the `protocol-jackson` module, where it already maps the value type in a single location to both the intent and the implementation classes.

The proposal above would mean that:

- You register the `ValueType` -> `Intent` mapping directly in the enum and only in one place
- You register the expected `RecordValue` interface type mapping directly in the enum, meaning that implementations can verify they are indeed deserializing the right thing. This has the added benefit that it's also explicit to anyone looking at the `ValueType` enum what the expected analogous `RecordValue` and `Intent` are, instead of being somehow informally specified in different modules.

**Comment by npepinpe** [comment]
After discussion, we will drop this for now - the benefits are marginal, and it hasn't been a particular pain point for the team afaik.
